classes/migration/upgrade/v3_4_0/PreflightCheckStatsMigration.inc.php

```diff
@@ -0,0 +1,179 @@
+            // Prevent further downgrade migrations from executing.
+            $this->_installer->migrations = [];
+        }
+    }
+
+    /**
+     * Store the fallback version in the database, permitting resumption of partial upgrades.
+     *
+     * @return ?string Fallback version, if one was identified
+     */
+    protected function setFallbackVersion(): ?string
+    {
+        if ($fallbackVersion = $this->_attributes['fallback'] ?? null) {
+            $versionDao = DAORegistry::getDAO('VersionDAO'); /** @var VersionDAO $versionDao */
+            $versionDao->insertVersion(\PKP\site\Version::fromString($fallbackVersion));
+            return $fallbackVersion;
+        }
+        return null;
+    }
+
+    /**
+     * Check if there are unprocessed usage event log files (older than today).
+     *
+     *  @throws \Exception
+     */
+    private function checkUsageEventLogFiles(): void
+    {
+        // Check if usageStats folders contain any log file other than the current one.
+        $processingDirPath = StatisticsHelper::getUsageStatsDirPath() . '/processing';
+        $processingDirFiles = glob($processingDirPath . '/' . '*');
+        $stageDirPath = StatisticsHelper::getUsageStatsDirPath() . '/stage';
+        $stageDirFiles = glob($stageDirPath . '/' . '*');
+        $rejectDirPath = StatisticsHelper::getUsageStatsDirPath() . '/reject';
+        $rejectDirFiles = glob($rejectDirPath . '/' . '*');
+        $usageEventLogsDirPath = StatisticsHelper::getUsageStatsDirPath() . '/usageEventLogs';
+        $usageEventLogsDirFiles = glob($usageEventLogsDirPath . '/' . '*');
+        $usageEventLogsDirFilesError = false;
+        if (is_array($usageEventLogsDirFiles)) {
+            $fileMgr = new FileManager();
+            foreach ($usageEventLogsDirFiles as $filePath) {
+                // Make sure it's a file.
+                if ($fileMgr->fileExists($filePath)) {
+                    // If file is not from today.
+                    $filename = pathinfo($filePath, PATHINFO_BASENAME);
+                    $currentDayFilename = 'usage_events_' . date('Ymd') . '.log';
+                    if ($filename != $currentDayFilename) {
+                        $usageEventLogsDirFilesError = true;
+                        break;
+                    }
+                }
+            }
+        }
+        if ((is_array($processingDirFiles) && count($processingDirFiles)) ||
+            (is_array($stageDirFiles) && count($stageDirFiles)) ||
+            (is_array($rejectDirFiles) && count($rejectDirFiles)) || $usageEventLogsDirFilesError) {
+            $msg = 'Please process all usage event log files older than today before running the upgrade. Check for log files in all usageStats folders (usageEventLogs, processing, stage and reject).';
```

Comment:
Yes, I was not sure about this check either. 
Maybe I should remove it. This would mean that those users will not have the stats for those days. Once when I implement the tool for log file reprocessing they will be able to reprocess them and will then get the stats. It would be better for them to do it immediately i.e. before upgrade, but... 
This check is also a reason why the travis tests are failing -- because there is at least one log file there... :-)

Reply:
I will remove this check for now...

Reply:
@asmecher what do you think? Should we block upgrades when log files exist that haven't been processed?

Reply:
Hmm, yeeeessss, this is going to be a pain and frequent topic of discussion on the forum but on balance I'd support asking the user to process logs before upgrading. This is a good use of the preflight check!

I would suggest making the message more detailed, though, including...
- How to process the stats from the command line and/or web interface (remember that many users will not have command line experience)
- How to get rid of the log files if they aren't important to them (e.g. for a test upgrade)

Reply:
I would maybe leave the decision for this preflight check for when I implemented the tool for the log file reprocessing, OK? For now I will remove it...
We can definitely add the information about the log file proessing somewhere where the users read how to upgrade to this version.
This preflight check is also problematic because the user will not have the old code any more, which is at the moment (till I haven't implemented the new tool) needed for the reprocessing of the old log files.